os/kernel/fault_manager/fault_manager.c

```diff
@@ -0,0 +1,285 @@
+ *
+ * Returned Value:
+ *   pid   -   pid of the root application of input pid
+ *
+ ****************************************************************************/
+pid_t get_appid(pid_t pid)
+{
+	/* TODO: Implementation of traverser of the tree to get root app pid */
+	return pid;
+}
+
+/****************************************************************************
+ * Name: kill_app
+ *
+ * Description:
+ *   This function will kill all the task/thread created by the root application
+ *   i.e input pid.
+ *   If any fault occurs in the system, fault manager will call this function to
+ *   kill all the thread/task created by the input application.
+ *
+ * Input parameters:
+ *   pid   -   The pid of the applciation, whoes all childs to be killed
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ ****************************************************************************/
+int kill_app(pid_t pid)
+{
+	int ret;
+
+	/* TODO: Implementation of traverser of tree and kill all the task/thread
+	 * of the root application
+	 */
+	ret = kill(pid, SIGWORK);
+	if (ret == -1) {
+		return ERROR;
+	}
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: fault_manager_init
+ *
+ * Description:
+ *   This function will receive the faulty application's pid and check if it is one
+ *   of the registered application for fault manager.
+ *   If the application is registered then it will be restarted after killing its
+ *   child processes, and if it is not registered then board will be rebooted.
+ *
+ * Input parameters:
+ *   None
+ *
+ * Returned Value:
+ *   Zero (OK) on success; otherwise -1 (ERROR) value is returned.
+ *
+ * Assumptions:
+ *    All the applications of the system will be registered for the fault manager
+ *
+ ****************************************************************************/
+int fault_manager_init(int argc, char *argv[])
+{
+	mqd_t mqdes;
+	int ret;
+	int i;
+	int index = -1;
+	pid_t pid;
+	pid_t pid_restart;
+	pid_t app_id;
+	struct mq_attr attr;
+	char restart_app[PATH_LEN] = " ";
+
+	attr.mq_maxmsg = 32;
+	attr.mq_msgsize = sizeof(pid);
+	attr.mq_flags = 0;
+
+	mqdes = mq_open("fault_queue", O_CREAT | O_RDONLY, 0666, &attr);
+	if (mqdes == (mqd_t) - 1) {
+		svdbg("Unable to open the queue\n");
+		return ERROR;
+	}
+
+	while (1) {
+		ret = mq_receive(mqdes, (char *)&pid, sizeof(pid), NULL);
+		if (ret == -1) {
+			svdbg("Unable to receive \n");
+			return ERROR;
+		}
+
+		app_id = get_appid(pid);
+
+		svdbg("faulty app pid = %d\n", app_id);
+
+		for (i = 0; i < CONFIG_NUM_APPS; i++) {
+			if (reg_table[i].appid == app_id) {
+				index = i;
+			}
+		}
+
+		if (index >= 0 && index < CONFIG_NUM_APPS) {
+			strcpy(restart_app, reg_table[index].app_path);
+
+			kill_app(app_id);
+
+			fm_deregister_app(reg_table[index].appid);
+			index = -1;
+			sleep(2);
+
+			if ((pid_restart = exec(restart_app, NULL, NULL, 0)) <= 0) {
```

Comment:
What if the app is partially linked elf?

Reply:
I have implemented this based on the example of sample elf apps, 
Once we have actual implementation of partially linked elf, I will modify accordingly.
I need to see how are we are planning to initiates the applications.